Add plugin manager for sources

mszostok opened the pull request:
## Description

Changes proposed in this pull request:

- Add plugin manager for sources
- Update contrib doc and provide an easy way to develop Botkube core plugins
- Support serving source via fake plugin HTTP server
- Add e2e tests

## Related issue(s)

- #838

mszostok commented on test/fake/plugin_server.go:
```diff
@@ -1,4 +1,4 @@
-package e2e
+package fake
```
I cannot put the `fake` pkg inside `e2e` as then logs streaming for test doesn't work because of this issue: https://github.com/golang/go/issues/46959

mszostok commented on CONTRIBUTING.md:
````diff
@@ -130,6 +130,44 @@ For faster development, you can also build and run Botkube outside K8s cluster.
+   go run test/helpers/plugin_server.go
+   ```
+
+	 > **Note**
+	 > If Botkube runs inside the k3d cluster, export the `PLUGIN_SERVER_HOST=http://host.k3d.internal` environment variable.
+
````
```suggestion
   > **Note**
   > If Botkube runs inside the k3d cluster, export the `PLUGIN_SERVER_HOST=http://host.k3d.internal` environment variable.

```

josefkarasek commented on cmd/source/cm-watcher/main.go:
```diff
@@ -0,0 +1,106 @@
+)
+
+// CMWatcher implements Botkube source plugin.
+type CMWatcher struct{}
```
Code style consistency - `type` block above combined with single line declaration.

mszostok replied:
Actually, So I use the `type` block to "group" related entities, so the code block about groups types related to the configuration. This one is about the "service" that implements the source plugin functionality. 

Are you OK with such approach?

josefkarasek commented on internal/source/source.go:
```diff
@@ -0,0 +1,184 @@
+	// As a result our key is e.g. ['source-name1;source-name2']
+	startProcesses := map[string]struct{}{}
+
+	startPlugin := func(bindSources []string) error {
```
Isn't this function complex enough to be named / defined separately?

mszostok replied:
yes, you are right. I changed it to become a dedicated method on this struct 👍 

In the next PR, I will change it even more to make it more straight forward when it comes to the responsibilities. Because I see that it is to complex to maintain and test all those pieces together.

josefkarasek reviewed (approved):
lgtm :+1:

mszostok commented:
Hi @josefkarasek!

Thanks for your review 🙇 
I addressed the feedback and also merge the PR as it was a dependency for my second PR. You were also assigned there 🙂 so we will be able to follow up there with some changes that you will see as required. Thanks!
